simulation/traffic_simulator/src/entity/ego_entity.cpp

```diff
@@ -161,6 +161,8 @@ EgoEntity::EgoEntity(
   vehicle_model_type_(getVehicleModelType()),
   vehicle_model_ptr_(makeSimulationModel(vehicle_model_type_, step_time, parameters))
 {
+  // TODO: this job should be activate but needs data obtained from Autoware in BehaviorParameter
+  job_list_.inactivate(traffic_simulator::job::Type::OUT_OF_RANGE_DYNAMIC_CONSTRAINS);
```

Comment:
Out of range check is necessary for EgoEntity.

Reply:
@hakuturu583 
I also believe that `out_of_range` should be activated for Ego. However, as described [here](https://github.com/tier4/scenario_simulator_v2/pull/979#discussion_r1145919231), because of [this](https://github.com/tier4/scenario_simulator_v2/blob/cf575c0fb9e9bbb6f62453fb4aa6fdf552b0eb4a/simulation/traffic_simulator/src/entity/ego_entity.cpp#L180-L190) method cannot be:
```
auto EgoEntity::getBehaviorParameter() const -> traffic_simulator_msgs::msg::BehaviorParameter
{
  traffic_simulator_msgs::msg::BehaviorParameter parameter;
  /**
   * @brief TODO, Input values get from autoware.
   */
  parameter.see_around = true;
  parameter.dynamic_constraints.max_acceleration = 0;
  parameter.dynamic_constraints.max_deceleration = 0;
  return parameter;
}
```
I mentioned it in a comment [here](https://tier4.atlassian.net/browse/AJD-696?focusedCommentId=117533) and suggested changes ([commit](https://github.com/tier4/scenario_simulator_v2/pull/979/commits/54f5abce604da9d679e207a82fe99671a4fb8964)):

> I'm not sure why there were assignments like max acceleration=0 in EgoEntity::getBehaviorParameter() - without commenting out [these 2 lines](https://github.com/tier4/scenario_simulator_v2/blob/c425a60edc321fcfd56d4404ec42f0a64865d3e4/simulation/traffic_simulator/src/entity/ego_entity.cpp#L185-L186), out_of_range would not allow Ego to move.
> 
> However, this method still returned msgs::BehaviourParameter - not taking Performance into account. That's why I’ve removed this method and the definition from [VehicleEntity](https://github.com/tier4/scenario_simulator_v2/blob/54f5abce604da9d679e207a82fe99671a4fb8964/simulation/traffic_simulator/src/entity/vehicle_entity.cpp#L74-L77) is used.
> 
> I'm not sure if Performance should be taken into account in these cases, thats why I asked. :slight_smile:

In summary: 
In order to run `out_of_range` for Ego it is necessary to change the definition of `EgoEntity::getBehaviorParameter()`. 
For this reason I removed the definitions of `EgoEntity::getBehaviorParameter()` and `EgoEntity::setBehaviorParameter()` - the definition from [VehicleEntity](https://github.com/tier4/scenario_simulator_v2/blob/54f5abce604da9d679e207a82fe99671a4fb8964/simulation/traffic_simulator/src/entity/vehicle_entity.cpp#L74-L77) was used. 
This allowed not only to activate the job for Ego, but also ensures that Performance limits were taken into account.


However, you answered:
> This line [( entity/ego_entity.cpp#L181-L187)](https://github.com/tier4/scenario_simulator_v2/blob/f9357649bce1b2990574086398061d3afe2cba37/simulation/traffic_simulator/src/entity/ego_entity.cpp#L181-L187) shows that the part is still WIP and should not be changed without checking and considering the circumstances of why this is happening.
> I don't think it is a good implementation to use a base type function without inputting data obtained from Autoware in BehaviorParameter in the first place.

So I thought that since [these lines](https://github.com/tier4/scenario_simulator_v2/blob/c425a60edc321fcfd56d4404ec42f0a64865d3e4/simulation/traffic_simulator/src/entity/ego_entity.cpp#L185-L186) _should not be changed_, reverting changes and deactivating `out_of_range` for Ego is the only solution. 

If indeed  `EgoEntity::getBehaviorParameter()` cannot be modified and this check is necessary for `EgoEntity`:
I suggest to redefine `addOutOfRangeJob()` in `EgoEntity` and there provide const limits from raw `traffic_simulator_msgs::msg::DynamicConstraints` message:
```
void EgoEntity::addOutOfRangeJob()
{
  /**
   * @brief This value was determined heuristically rather than for
   * technical reasons.
   */
  constexpr double tolerance = 0.01;
  job_list_.append(
    /**
     * @brief Checking if the values of velocity, acceleration and jerk are within the acceptable
     * range
     */
    [this, tolerance](double) {
      auto velocity_ = status_.action_status.twist.linear.x;
      auto accel_ = status_.action_status.accel.linear.x;
      auto jerk_ = status_.action_status.linear_jerk;
      auto d_c = traffic_simulator_msgs::msg::DynamicConstraints;
      if (std::abs(velocity_) - d_c.max_speed > tolerance)
        THROW_SPECIFICATION_VIOLATION(
          "Entity: ", name, " - current velocity (which is ", velocity_,
          ") is out of range (which is [", -d_c.max_speed, ", ", d_c.max_speed, "])");

      if (accel_ - d_c.max_acceleration > tolerance || -accel_ - d_c.max_deceleration > tolerance)
        THROW_SPECIFICATION_VIOLATION(
          "Entity: ", name, " - current acceleration (which is ", accel_,
          ") is out of range (which is [", -d_c.max_deceleration, ", ", d_c.max_acceleration, "])");

      if (std::abs(jerk_) - max_jerk_ > tolerance)
        THROW_SPECIFICATION_VIOLATION(
          "Entity: ", name, " - current jerk (which is ", jerk_, ") is out of range (which is [",
          -max_jerk_, ", ", max_jerk_, "])");
      return false;
    },
    /**
     * @brief This job is always ACTIVE
     */
    [this]() {}, job::Type::OUT_OF_RANGE_DYNAMIC_CONSTRAINS, true, job::Event::POST_UPDATE);
}
```

Reply:
OutOfRangeMetric is a metric to verify if the velocity, etc. is within the specified range, and we have confirmed that there is no need to follow if DynamicConstriaints, etc. are subsequently updated.
As with the current implementation of openscenario_interpretor, after calling API::spawn function, API::checkOutofRange(const std::string & entity_name, speed_limit, jerk_limit, etc...) function to initiate the check.

Translated with www.DeepL.com/Translator (free version)